File: automation-service/routes/exec.py
"""
AI Execution Proxy Router
Proxies execution requests to ai-pipeline service
"""
from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field, validator
from typing import Optional, Dict, Any
import logging
import time
import uuid

from clients.ai_pipeline import (
    AIPipelineClient,
    AIPipelineTimeoutError,
    AIPipelineBadGatewayError,
    AIPipelineError
)
from config import config
from observability.metrics import (
    record_ai_request_success,
    record_ai_request_error
)


logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ai", tags=["AI Execution"])


# ============================================================================
# REQUEST/RESPONSE MODELS
# ============================================================================

class ExecRequest(BaseModel):
    """Request to execute a tool via AI Pipeline"""
    input: str = Field(..., min_length=1, max_length=4000, description="Input text for the tool")
    tool: Optional[str] = Field(default="echo", description="Tool to execute (default: echo)")
    trace_id: Optional[str] = Field(default=None, description="Optional trace ID for request tracking")
    
    @validator('input')
    def validate_input(cls, v):
        if not v or not v.strip():
            raise ValueError("Input cannot be empty")
        return v


class ErrorDetail(BaseModel):
    """Error detail structure"""
    code: str
    message: str


class ExecResponse(BaseModel):
    """Response from AI execution"""
    success: bool
    output: Optional[str] = None
    error: Optional[ErrorDetail] = None
    trace_id: str
    duration_ms: float
    tool: str


# ============================================================================
# GLOBAL CLIENT INSTANCE
# ============================================================================

_ai_pipeline_client: Optional[AIPipelineClient] = None


def get_ai_pipeline_client() -> AIPipelineClient:
    """Get or create AI Pipeline client singleton"""
    global _ai_pipeline_client
    
    if _ai_pipeline_client is None:
        _ai_pipeline_client = AIPipelineClient(
            base_url=config.AI_PIPELINE_BASE_URL,
            timeout_seconds=config.get_exec_timeout_seconds()
        )
        logger.info(f"[Exec] AI Pipeline client initialized: {config.AI_PIPELINE_BASE_URL}")
    
    return _ai_pipeline_client


# ============================================================================
# ROUTES
# ============================================================================

@router.post("/execute", response_model=ExecResponse)
async def execute_tool(
    request: ExecRequest,
    x_trace_id: Optional[str] = Header(None, alias="X-Trace-Id")
):
    """
    Execute a tool via AI Pipeline proxy
    
    This endpoint proxies execution requests to the ai-pipeline service,
    providing a walking skeleton for end-to-end execution testing.
    
    Args:
        request: Execution request with input and tool
        x_trace_id: Optional trace ID from header
    
    Returns:
        ExecResponse with execution result
    
    Raises:
        HTTPException: On validation or execution errors
    """
    start_time = time.perf_counter()
    
    # Determine trace_id (header > body > generated)
    trace_id = x_trace_id or request.trace_id or str(uuid.uuid4())
    tool = request.tool or "echo"
    
    logger.info(f"[Exec] start input=\"{request.input[:80]}\" tool={tool} trace_id={trace_id}")
    
    try:
        # Get AI Pipeline client
        client = get_ai_pipeline_client()
        
        # Call ai-pipeline /execute
        result = await client.execute(
            tool=tool,
            input_text=request.input,
            trace_id=trace_id
        )
        
        # Calculate duration
        duration_ms = (time.perf_counter() - start_time) * 1000
        
        # Check if ai-pipeline reported success
        if result.get("success"):
            # Record success metrics
            record_ai_request_success(tool, duration_ms / 1000)
            
            logger.info(
                f"[Exec] done duration={duration_ms:.2f}ms success=True trace_id={trace_id}"
            )
            
            return ExecResponse(
                success=True,
                output=result.get("output"),
                trace_id=trace_id,
                duration_ms=round(duration_ms, 2),
                tool=tool
            )
        else:
            # AI Pipeline returned success=false
            error_msg = result.get("error", "Unknown error from AI Pipeline")
            
            # Record error metrics
            record_ai_request_error(tool, "upstream_error", duration_ms / 1000)
            
            logger.warning(
                f"[Exec] done duration={duration_ms:.2f}ms success=False trace_id={trace_id} "
                f"error={error_msg}"
            )
            
            return ExecResponse(
                success=False,
                error=ErrorDetail(code="upstream_error", message=error_msg),
                trace_id=trace_id,
                duration_ms=round(duration_ms, 2),
                tool=tool
            )
    
    except AIPipelineTimeoutError as e:
        duration_ms = (time.perf_counter() - start_time) * 1000
        
        # Record timeout metrics
        record_ai_request_error(tool, "timeout", duration_ms / 1000)
        
        logger.error(f"[Exec] timeout trace_id={trace_id} error={e}")
        
        raise HTTPException(
            status_code=502,
            detail={
                "success": False,
                "error": {
                    "code": "timeout",
                    "message": str(e)
                },
                "trace_id": trace_id,
                "duration_ms": round(duration_ms, 2),
                "tool": tool
            }
        )
    
    except AIPipelineBadGatewayError as e:
        duration_ms = (time.perf_counter() - start_time) * 1000
        
        # Record bad gateway metrics
        record_ai_request_error(tool, "bad_gateway", duration_ms / 1000)
        
        logger.error(f"[Exec] bad_gateway trace_id={trace_id} error={e}")
        
        raise HTTPException(
            status_code=502,
            detail={
                "success": False,
                "error": {
                    "code": "bad_gateway",
                    "message": str(e)
                },
                "trace_id": trace_id,
                "duration_ms": round(duration_ms, 2),
                "tool": tool
            }
        )
    
    except AIPipelineError as e:
        duration_ms = (time.perf_counter() - start_time) * 1000
        
        # Record unknown error metrics
        record_ai_request_error(tool, "unknown", duration_ms / 1000)
        
        logger.error(f"[Exec] error trace_id={trace_id} error={e}")
        
        raise HTTPException(
            status_code=502,
            detail={
                "success": False,
                "error": {
                    "code": "unknown",
                    "message": str(e)
                },
                "trace_id": trace_id,
                "duration_ms": round(duration_ms, 2),
                "tool": tool
            }
        )
    
    except Exception as e:
        duration_ms = (time.perf_counter() - start_time) * 1000
        
        # Record validation/unknown error
        record_ai_request_error(tool, "validation", duration_ms / 1000)
        
        logger.error(f"[Exec] unexpected_error trace_id={trace_id} error={e}")
        
        raise HTTPException(
            status_code=400,
            detail={
                "success": False,
                "error": {
                    "code": "validation",
                    "message": str(e)
                },
                "trace_id": trace_id,
                "duration_ms": round(duration_ms, 2),
                "tool": tool
            }
        )

File: frontend/src/services/exec.ts
/**
 * Exec Service
 * Service for calling automation-service /ai/execute endpoint
 */
import axios from 'axios';

// Get automation service base URL
const getAutomationServiceUrl = (): string => {
  // Use environment variable or default to localhost:8010
  return process.env.REACT_APP_AUTOMATION_SERVICE_URL || 'http://localhost:8010';
};

// Get exec base path
const getExecBasePath = (): string => {
  return process.env.REACT_APP_EXEC_BASE_PATH || '/ai/execute';
};

// Check if exec sandbox is enabled
export const isExecEnabled = (): boolean => {
  return process.env.REACT_APP_EXEC_ENABLE === 'true';
};

/**
 * Exec request interface
 */
export interface ExecRequest {
  input: string;
  tool?: string;
  trace_id?: string;
}

/**
 * Error detail from backend
 */
export interface ErrorDetail {
  code: string;
  message: string;
}

/**
 * Exec response interface
 */
export interface ExecResponse {
  success: boolean;
  output?: string;
  error?: ErrorDetail;
  trace_id: string;
  duration_ms: number;
  tool: string;
}

/**
 * Execute a tool via automation-service proxy
 * 
 * @param input - Input text for the tool
 * @param tool - Tool name (default: "echo")
 * @param traceId - Optional trace ID for request tracking
 * @returns Promise<ExecResponse>
 */
export async function execRun(
  input: string,
  tool: string = 'echo',
  traceId?: string
): Promise<ExecResponse> {
  const url = `${getAutomationServiceUrl()}${getExecBasePath()}`;
  
  console.log(`[Exec] start input="${input.substring(0, 80)}" tool=${tool}`);
  
  const startTime = performance.now();
  
  try {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    
    // Add trace ID to header if provided
    if (traceId) {
      headers['X-Trace-Id'] = traceId;
    }
    
    const response = await axios.post<ExecResponse>(
      url,
      {
        input,
        tool,
        trace_id: traceId,
      },
      { headers }
    );
    
    const duration = performance.now() - startTime;
    
    console.log(
      `[Exec] done duration=${duration.toFixed(2)}ms success=${response.data.success} trace_id=${response.data.trace_id}`
    );
    
    return response.data;
  } catch (error: any) {
    const duration = performance.now() - startTime;
    
    // Handle HTTP error responses
    if (error.response?.data) {
      const errorData = error.response.data;
      
      console.error(
        `[Exec] error duration=${duration.toFixed(2)}ms status=${error.response.status} trace_id=${errorData.trace_id || 'unknown'}`
      );
      
      // If backend returned structured error, use it
      if (errorData.error) {
        return {
          success: false,
          error: errorData.error,
          trace_id: errorData.trace_id || 'unknown',
          duration_ms: errorData.duration_ms || duration,
          tool: errorData.tool || tool,
        };
      }
    }
    
    // Generic error handling
    console.error(`[Exec] error duration=${duration.toFixed(2)}ms`, error);
    
    return {
      success: false,
      error: {
        code: 'client_error',
        message: error.message || 'Unknown error occurred',
      },
      trace_id: 'unknown',
      duration_ms: duration,
      tool,
    };
  }
}